README.md

```diff
@@ -51,17 +51,17 @@ In witness automata, the GraphML edges represent state transitions.
+| endoffset | *Valid values:* Offset of a specific character in the program from the program start. <br /> ``endoffset`` is the same as the ``startoffset`` key, except that it refers to the character offset at the end of an operation. | Yes | Yes |
+| enterLoopHead | *Valid values:* ``false`` (default) or ``true`` <br /> ``enterLoopHead`` is used as part of the source-code guard of a transition and restricts the set of CFA edges matched by the source-code guard to operations on CFA edges where the successor is a loop head. For our format specification, any CFA node that (1) is part of a loop in the CFA, (2) has an entering CFA edge where the predecessor node is not in the loop, and (3) has a leaving CFA edge where the successor node is not in the loop, qualifies as a loop head. Note, however, that depending on the programming language of the verification task, the loop head may be ambiguous. For example, in C it is possible to use ``goto`` statements to construct arbitrarily complex loops with many CFA nodes that match the definition above. The ``attr.type`` attribute of this key is ``boolean`` and its default value is ``false``. | Yes | Yes |
+| enterFunction | *Valid values:* Function name <br /> ``enterFunction`` is used as part of the source-code guard of a transition and restricts the set of CFA edges matched by the source-code guard to function-call operations where the name of the called function matches the specified value. The ``attr.type`` attribute of this key is ``string``. A witness consumer may also use this key to track a stack of function calls and use this information to qualify ambiguously named variables in state-space guards or state invariants in the absence of explicitly specified scopes via the ``assumption.scope`` or ``invariant.scope`` keys. The ``attr.type`` attribute of this key is ``string``. | Yes | Yes |
+| returnFromFunction | *Valid values:* Function name <br /> ``returnFromFunction`` is the counterpart of the ``enterFunction`` key, i.e., it is used as part of the source-code guard of a transition and restricts the set of CFA edges matched by the source-code guard to function-return operations where the name of the function that is being returned from matches the specified value. Analogously to ``enterFunction``, a witness consumer may use this key to track a stack of function calls. The ``attr.type`` attribute of this key is ``string``. | Yes | Yes |
```

Comment:
This touches issue #36 and clarifies that there is no need for a matching `returnFromFunction`, correct?

The question with callstack for parallel programs is resolved to imho, since we do not pose any constraints here on how the callstack that the witness consumer constructs is structured, so it is up to the witness consumer to construct a callstack if their liking (especially one that is per thread).

@kfriedberger do you agree? Any comments?

Reply:
> This touches issue #36 and clarifies that there is no need for a matching returnFromFunction, correct?

No, that was not my intention. In fact, "a witness consumer may use this key to track a stack of function calls" would not work if there is no matching returnFromFunction in a run through the automaton. Note that a control state of a witness automaton may have multiple leaving transitions with different returnFromFunctions; the validator should then proceed along those transitions that make sense regarding its tracked callstack and stutter otherwise.

Suggestions for improvements to the wording to clarify are much appreciated.

Reply:
For parallel programs, the transition should already contain a `threadId` and thus match a corresponding thread.
If the `threadId` is not available, any thread can apply the given function call or return.
This seems to be valid and might not need rewording.